src/os/tempfile.go

```diff
@@ -95,6 +95,48 @@ func prefixAndSuffix(pattern string) (prefix, suffix string, err error) {
 	return prefix, suffix, nil
 }
 
+// MkdirTemp creates a new temporary directory in the directory dir
+// and returns the pathname of the new directory.
+// The new directory's name is generated by adding a random string to the end of pattern.
+// If pattern includes a "*", the random string replaces the last "*" instead.
+// If dir is the empty string, MkdirTemp uses the default directory for temporary files, as returned by TempDir.
+// Multiple programs or goroutines calling MkdirTemp simultaneously will not choose the same directory.
+// It is the caller's responsibility to remove the directory when it is no longer needed.
+func MkdirTemp(dir, pattern string) (string, error) {
+	if dir == "" {
+		dir = TempDir()
+	}
+
+	prefix, suffix, err := prefixAndSuffix(pattern)
+	if err != nil {
+		return "", &PathError{Op: "mkdirtemp", Path: pattern, Err: err}
+	}
+	prefix = joinPath(dir, prefix)
+
+	try := 0
+	for {
+		name := prefix + nextRandom() + suffix
+		err := Mkdir(name, 0700)
+		if err == nil {
+			return name, nil
+		}
+		if IsExist(err) {
+			if try++; try < 10000 {
+				continue
+			}
+			return "", &PathError{Op: "mkdirtemp", Path: dir + string(PathSeparator) + prefix + "*" + suffix, Err: ErrExist}
+		}
+		if IsNotExist(err) {
+			_, serr := Stat(dir)
+			ine := IsNotExist(err)
```

Comment:
Should this be checking `serr` instead of `err` ?

Reply:
thanks.  That is partially reverted (and evidently broken) debugging code.  Fixing...

Reply:
Squashed this fix in:

```
 		if IsNotExist(err) {
-			_, serr := Stat(dir)
-			ine := IsNotExist(err)
-			if ine {
-				return "", serr
+			if _, err := Stat(dir); IsNotExist(err) {
+				return "", err
 			}
 		}
```

to make it mirror the upstream test as intended.